plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php

```diff
@@ -389,17 +528,31 @@ public function get_total_pending_count(): int {
 	 * @return array Batch of records.
 	 */
 	public function get_next_batch_to_process( int $size ): array {
-		if ( $this->custom_orders_table_is_authoritative() ) {
-			$order_ids = $this->get_ids_of_orders_pending_sync( self::ID_TYPE_MISSING_IN_POSTS_TABLE, $size );
-		} else {
-			$order_ids = $this->get_ids_of_orders_pending_sync( self::ID_TYPE_MISSING_IN_ORDERS_TABLE, $size );
-		}
+		$orders_table_is_authoritative = $this->custom_orders_table_is_authoritative();
+
+		$order_ids = $this->get_ids_of_orders_pending_sync(
+			$orders_table_is_authoritative ? self::ID_TYPE_MISSING_IN_POSTS_TABLE : self::ID_TYPE_MISSING_IN_ORDERS_TABLE,
+			$size
+		);
 		if ( count( $order_ids ) >= $size ) {
 			return $order_ids;
 		}
 
 		$order_ids = $order_ids + $this->get_ids_of_orders_pending_sync( self::ID_TYPE_DIFFERENT_UPDATE_DATE, $size - count( $order_ids ) );
-		return $order_ids;
+		if ( count( $order_ids ) > 0 ) {
+			return $order_ids;
+		}
+
+		// Deleted orders pending sync have the lowest priority, and additionally,
+		// if such orders exist then we return just those, even if they don't fill a batch.
```

Comment:
Is there a specific reason for this? Because otherwise, this will interfere with time remaining etc calculation in CLI (not that it's super accurate in the first place).

Reply:
The reason is the second part of the comment: "Returning 'missing orders that are to be deleted' and 'missing orders that are to be created' in the same batch is confusing and potentially dangerous". But isn't CLI using `get_current_orders_pending_sync_count` or something similar for displaying progress information?

Reply:
> in the same batch is confusing and potentially dangerous
Looks like we have a check in the delete function already to look for `_delete_from` metadata, would you still say this could be dangerous? My reason for suggesting keeping the batch size the same is to be consistent with the current behavior, which can have orders to be inserted and to be updated as part of the same batch.

> get_current_orders_pending_sync_count
It does, but I think the percentage calculator assumes batch to have same size.

Reply:
Ok, `process_batch` is already telling apart deletion records from actual updates, so I guess we can change this part.

Reply:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/91d5546cea1f57216770e9ff22ae30b5afca46af